nilearn/decoding/space_net.py

```diff
@@ -0,0 +1,1231 @@
+
+    return X, mask_, support
+
+
+def _space_net_alpha_grid(X, y, eps=1e-3, n_alphas=10, l1_ratio=1.,
+        logistic=False):
+    """Compute the grid of alpha values for TV-L1 and S-Lasso.
+
+    Parameters
+    ----------
+    X : ndarray, shape (n_samples, n_features)
+        Training data (design matrix).
+
+    y : ndarray, shape (n_samples,)
+        Target / response vector.
+
+    l1_ratio : float
+        The ElasticNet mixing parameter, with ``0 <= l1_ratio <= 1``.
+        For ``l1_ratio = 0`` the penalty is purely a spatial prior
+        (S-LASSO, TV, etc.). ``For l1_ratio = 1`` it is an L1 penalty.
+        For ``0 < l1_ratio < 1``, the penalty is a combination of L1
+        and a spatial prior
+
+    eps : float, optional
+        Length of the path. ``eps=1e-3`` means that
+        ``alpha_min / alpha_max = 1e-3``
+
+    n_alphas : int, optional
+        Number of alphas along the regularization path.
+
+    """
+
+    if logistic:
+        # Computes the theoretical upper bound for the overall
+        # regularization, as derived in "An Interior-Point Method for
+        # Large-Scale l1-Regularized Logistic Regression", by Koh, Kim,
+        # Boyd, in Journal of Machine Learning Research, 8:1519-1555,
+        # July 2007.
+        # url: http://www.stanford.edu/~boyd/papers/pdf/l1_logistic_reg.pdf
+        m = float(y.size)
+        m_plus = float(y[y == 1].size)
+        m_minus = float(y[y == -1].size)
+        b = np.zeros_like(y)
+        b[y == 1] = m_minus / m
+        b[y == -1] = - m_plus / m
+        alpha_max = np.max(np.abs(X.T.dot(b)))
+
+        # tt may happen that b is in the kernel of X.T!
+        if alpha_max == 0.:
+            alpha_max = np.abs(np.dot(X.T, y)).max()
+    else:
+        alpha_max = np.abs(np.dot(X.T, y)).max()
+
+    # prevent alpha_max from exploding when l1_ratio = 0
+    if l1_ratio == 0.:
+        l1_ratio = 1e-3
+    alpha_max /= l1_ratio
+
+    if n_alphas == 1:
+        return np.array([alpha_max])
+
+    alpha_min = alpha_max * eps
+    return np.logspace(np.log10(alpha_min), np.log10(alpha_max),
+                       num=n_alphas)[::-1]
+
+
+class EarlyStoppingCallback(object):
+    """Out-of-bag early stopping
+
+        A callable that returns True when the test error starts
+        rising. We use a Spearman correlation (btween X_test.w and y_test)
+        for scoring.
+    """
+    def __init__(self, X_test, y_test, is_classif, debias=False, verbose=0):
+        self.X_test = X_test
+        self.y_test = y_test
+        self.is_classif = is_classif
+        self.debias = debias
+        self.verbose = verbose
+        self.tol = -1e-4 if self.is_classif else -1e-2
+        self.test_scores = []
+        self.counter = 0.
+
+    def __call__(self, variables):
+        """The callback proper """
+        # misc
+        if not isinstance(variables, dict):
+            variables = dict(w=variables)
+        self.counter += 1
+        if self.counter == 0:
```

Comment:
Sorry, I don't see when this happens as self.counter is initialized to 0 and incremented before the test.